doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,617 @@
+      inapplicable to QUIC, so it is not supported.
+
+    - **C:** Not handshake-layer related. QUIC connection SSL object usage only.
+      Fails on a QUIC stream SSL object.
+
+    - **CS:** Not handshake-layer related. Can be used on any QUIC SSL object.
+
+### Existing APIs
+
+#### `SSL_set_connect_state`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+#### `SSL_set_accept_state`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+**Note:** Attempting to proceed in this state will not function for now because
+we do not implement server support at this time. However, the semantics of this
+function as such are unchanged.
+
+#### `SSL_is_server`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | HL            |
+
+#### `SSL_connect`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+Simple composition of `SSL_set_connect_state` and `SSL_do_handshake`.
+
+#### `SSL_accept`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+Simple composition of `SSL_set_accept_state` and `SSL_do_handshake`.
+
+#### `SSL_do_handshake`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | HL            |
+
+**Note:** Idempotent if handshake already completed.
+
+**Blocking Considerations:** Blocks until handshake completed if in blocking
+mode.
+
+**TBD:** Should this wait until handshake is completed or until it is confirmed?
+
+#### `SSL_read`, `SSL_read_ex`, `SSL_peek`, `SSL_peek_ex`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+**Blocking Considerations:** Blocks until at least one byte is available or an
+error occurs if in blocking mode (including the peek functions).
+
+#### `SSL_write`, `SSL_write_ex`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Error/Want    | Yes       | CS            |
+
+We have to implement all of the following modes:
+
+- `SSL_ENABLE_PARTIAL_WRITE` on or off
+- `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER` on or off
+- Blocking mode on or off
+
+**Blocking Considerations:** Blocks until all data is written or an error occurs
+if in blocking mode.
```

Comment:
Does SSL_write() with TLS ensure that the application data is delivered to the other side? IMO, no. IMO we have some leeway on how concretely we handle this with QUIC. We can block SSL_write() only based on ability to transmit the QUIC dgrams. The question is how to handle flow control limits in regards to SSL_write() blocking.

Reply:
I'd suggest we only block until the data is accepted into our send stream buffers (and those buffers should have some upper limit regarding how big they are). Flow control will govern how quickly the buffers are emptied. So this will naturally exert a backwards pressure on the application to slow down its send rate to match the rate the receiving peer is able to handle.

Reply:
The points we can wait until we return from a blocking call seem to be:
- We have processed all the data and it's in some internal queue. We might attempt to send it to the kernel.
- We have sent all the packets to the kernel (once)
- All the data has been ACKed.

The last one does not make sense to me. It might have arrived on the other side, but the other application might not have seen it. If you want to wait until the application has processed it, it needs to send something back. This can be either something at the application level, or with a shutdown of the connection.

The 2nd point also doesn't really make much sense for me. There is no advantage to the application to wait longer.

The only thing that makes sense to me is that we have processed the data and the application can now remove or reuse those buffer.

Reply:
Sorry, ambiguous wording. The usual semantics are preserved. Reworded:

>Blocks until libssl has accepted responsibility for (i.e., copied) all data provided, or an error occurs, if in blocking mode.